### PR TITLE
[inductor] Fix reference_as_float gradcheck

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106581
* #106747
* #106870
* __->__ #106626

When `reference_as_float` is true, reference gradients will not have the same
dtype as the actual computed gradients. This fixes the issue by downcasting
before doing the comparison.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -386,9 +390,16 @@ def run(*ex, **kwargs):
...
+            if reference_in_float:
+                flat_expect = reference_to_expect(flat_actual, flat_grads)
+            else:
+                flat_expect = flat_grads
+
             self.assertEqual(
-                actual_grad,
-                correct_grad,
+                flat_actual,
+                flat_expect,
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:402]: In the forward they unflatten it before passing it to assertEqual, perhasp you should do here the same?

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -386,9 +389,15 @@ def run(*ex, **kwargs):
             self.assertEqual(len(results_that_require_grad), 0)
         else:
             actual_grad = compute_grads(example_inputs, kwargs, actual, grads)
+
+            if reference_in_float:
+                expect_grad = reference_to_expect(actual_grad, correct_grad)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:394]: why is it correct not to flatten/unflatten here? Don't the gradients have the same structure as the forward operation?

Comment (peterbell10) [reply test/inductor/test_torchinductor.py:394]: `compute_grads` returns from `torch.autograd.grad(...)` which is always a flat tuple according to the [docstring](https://pytorch.org/docs/stable/generated/torch.autograd.grad.html).

**Review by lezcano — APPROVED** [review]
Fair enough

**Pull request closed without merging**